src/activity_actor.cpp

```diff
@@ -1124,36 +1124,101 @@ std::unique_ptr<activity_actor> hotwire_car_activity_actor::deserialize( JsonVal
 
-        moves_before_quality = to_moves<int>( furn_type->hacksaw->duration() );
+        if( finish ) {
+            const furn_str_id new_furn = std::get<furn_str_id>( furn_tool_ad->result() );
+            if( !new_furn.is_valid() ) {
+                if( !testing ) {
+                    debugmsg( "%s furniture: %s invalid furniture", tool_name, new_furn.str() );
+                }
+                act.set_to_null();
+                return false;
+            }
+
+            data = &*furn_tool_ad;
+            here.furn_set( target, new_furn );
+        } else {
+            moves = to_moves<int>( ( *furn_type.*tool_member )->duration() );
+        }
     } else if( const ter_id &ter_type = here.ter( target ); !ter_type->is_null() ) {
-        if( !ter_type->hacksaw->valid() ) {
+        const auto ter_tool_ad = ( *ter_type.*tool_member );
+        if( !ter_tool_ad->valid() ) {
             if( !testing ) {
-                debugmsg( "%s hacksaw is invalid", ter_type.id().str() );
+                debugmsg( "%s %s is invalid", ter_type.id().str(), tool_name );
             }
             act.set_to_null();
-            return;
+            return false;
+        }
+        if( finish ) {
+            const ter_str_id new_ter = std::get<ter_str_id>( ter_tool_ad->result() );
+            if( !new_ter.is_valid() ) {
+                if( !testing ) {
+                    debugmsg( "%s terrain: %s invalid terrain", tool_name, new_ter.str() );
+                }
+                act.set_to_null();
+                return false;
+            }
+
+            data = &*ter_tool_ad;
+            here.ter_set( target, new_ter );
+        } else {
+            moves = to_moves<int>( ter_tool_ad->duration() );
         }
-        moves_before_quality = to_moves<int>( ter_type->hacksaw->duration() );
     } else {
         if( !testing ) {
-            debugmsg( "hacksaw activity called on invalid terrain" );
+            debugmsg( "%s activity called on invalid terrain", tool_name );
         }
         act.set_to_null();
+        return false;
+    }
+
+    if( finish ) {
+        for( const activity_byproduct &byproduct : data->byproducts() ) {
+            const int amount = byproduct.roll();
+            if( byproduct.item->count_by_charges() ) {
+                item byproduct_item( byproduct.item, calendar::turn, amount );
+                here.add_item_or_charges( target, byproduct_item );
+            } else {
+                item byproduct_item( byproduct.item, calendar::turn );
+                for( int i = 0; i < amount; ++i ) {
+                    here.add_item_or_charges( target, byproduct_item );
+                }
+            }
+        }
+
+        if( !data->message().empty() ) {
+            who.add_msg_if_player( m_info, data->message().translated() );
+        }
+
+        act.set_to_null();
+    }
+
+    return true;
+}
+
+void hacksaw_activity_actor::start( player_activity &act, Character &who )
+{
+    map &here = get_map();
```

Comment:
Using a single function for both start and finish means start can't pass in a `const map&` even though that code path won't modify it. I couldn't come up with an alternative that I considered legible.